### PR TITLE
Bug/async super

### PR DESCRIPTION
**Pull request by JGets**
https://trello.com/c/0SLNnpAC

- Fixes an issue where nested asynchronous `super` calls can result in an infinite loop

### CODE REVIEW

**Comment by JGets** [comment]
[(2) JS.js `super` creates infinite loop for nested calls on async methods](https://trello.com/c/0SLNnpAC/8297-2-jsjs-super-creates-infinite-loop-for-nested-calls-on-async-methods)

File: JS.js [hunk 1]
```diff
@@ -1132,17 +1126,36 @@ function clone(obj) {
 JS.util.clone = clone;
 
 /**
- * Helper function that runs a try-finally in a seperate function to allow compiler optimization
- * @param  {function} tryBlock     A function to be run in the `try` block
- * @param  {function} finallyBlock A function to be run in the `finally` block
+ * Helper function that runs two functions in a try-finally block, but will only run the `finallyBlock`
+ * function after any promises returned from the `tryBlock` function has fulfilled/rejected.
+ * The `finallyBlock` function will only ever be invoked once.
+ * @param  {Function} tryBlock       A (possibly asyncrhonous) function to be run in the `try` block
```
Comment (srmcconomy) [inline JS.js]: spelling on `asynchronous`

File: JS.js [hunk 2]
```diff
@@ -1132,17 +1126,36 @@ function clone(obj) {
...
+ * @param  {Function} [finallyBlock] A (synchronous) function to be run after the `tryBlock` function has executed and any returned promise has resolved
```
Comment (srmcconomy) [inline JS.js]: why aren't we allowing the finally block to be asynchronous?

File: test/super.js [hunk 3]
```diff
@@ -202,6 +247,34 @@ describe('super.js', function() {
...
+		const s1 = new Subclass();
+		s1.a   = 'Foo';
+		expect(await s1.asyncRecurse2(0)).to.equal('Sub-Base Foo');
+		expect(await s1.asyncRecurse2(1)).to.equal('Sub-Base Sub-Base Foo Foo');
+		expect(await s1.asyncRecurse2(2)).to.equal('Sub-Base Sub-Base Sub-Base Foo Foo Foo');
+		expect(await s1.asyncRecurse2(3)).to.equal('Sub-Base Sub-Base Sub-Base Sub-Base Foo Foo Foo Foo');
+
+
+		const s2 = new Subsubclass();
+		s2.a   = 'Bar';
```
Comment (srmcconomy) [inline test/super.js]: align

File: test/super.js [hunk 4]
```diff
@@ -202,6 +247,34 @@ describe('super.js', function() {
...
+
+		await subsubclass.asyncMethod('testing');
+		expect(subsubclass.a).to.equal('testing BaseClass');
+	});
+
+	it('should support cyclical recursive asynchronous super method calls', async function() {
+		this.timeout(5000);
+
+		const s1 = new Subclass();
+		s1.a   = 'Foo';
```
Comment (srmcconomy) [inline test/super.js]: align

File: JS.js [hunk 5]
```diff
@@ -712,19 +712,13 @@ function makeSuperStackUpdaterProxy(method) {
...
-
-			return tryFinally(
+			return possiblyAsyncTryFinally(
 				() => method.apply(this, args),
 				() => { instanceStack.pop(); proxy.superStack.set(this, instanceStack); }
 			);
 		}
 
-		return method.apply(this, arguments);
+		return method.apply(this, args);
```
Comment (srmcconomy) [inline JS.js]: I wonder if there is any difference between
```js
return method.apply(this, args);
```
,
```js
return method.call(this, ...args);
```
and
```js
return method(...args);
```

Comment (srmcconomy) [reply JS.js]: jsperf seems to indicate no difference, but I think option 3 is cleanest

File: JS.js (hunk 2 above)
Comment (JGets) [reply JS.js]: Mostly because we don't currently need that functionality, and adding it would introduce more complicated logic.

File: JS.js (hunk 1 above)
Comment (JGets) [reply JS.js]: 39f2f02

File: test/super.js (hunk 3 above)
Comment (JGets) [reply test/super.js]: 39f2f02

File: test/super.js (hunk 4 above)
Comment (JGets) [reply test/super.js]: 39f2f02

File: JS.js (hunk 5 above)
Comment (srmcconomy) [reply JS.js]: `this` is bound by default, is it not? Since `method` is not a member of an object

Comment (srmcconomy) [reply JS.js]:
```js
a = { f() { return this } };
a.f(); // { f: f }
function caller(method) { return method(); }
caller(a.f); // Window { ... }
```

Comment (JGets) [reply JS.js]: In [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#Securing_JavaScript), if unspecified, like for an anonymous function, `this` will be `undefined`:

```javascript
'strict mode';
a = { f() { return this } };
a.f(); // { f: f }
function caller(method) { return method(); }
caller(a.f); // undefined
```

Even out of strict mode, the method's `this` wouldn't necessarily be bound to the actual instance of the JSClass, so I think we would still need to `.apply` or `.call` it.

**Pull request merged**